### ISSUE TITLE
Device - Host - Device synchronization with VkEvent

### ISSUE THREAD

**Issue by neXyon** [KhronosGroup/Vulkan-Docs#755]
Shouldn't this be possible according to the specification? Here is a code sample:

https://gist.github.com/neXyon/859b2e52bac9a5a56b804d8a9d5fa4a5

I'm trying to synchronize a host stage into my pipeline, where I basically edit some data on the host during the execution of a command buffer on the device. From reading the specification I think I'm doing the correct synchronization, execution/memory dependencies and availability/visibility operations, but it neither works on NV nor AMD hardware. It works on Adreno though. I produced a minimal example (see link above) that shows how I'm trying to implement it:

1. A buffer is filled with increasing integers and uploaded to the device (device local buffer).
2. This buffer is copied to a host visible one.
3. The first event is set.
4. The second event is waited for.
5. Meanwhile the host waits for the first event.
6. After it has been set, it increments the numbers in the host visible buffer.
7. Then it sets the second event.
8.  The device then continues to copy the host visible buffer back to the device local buffer.
9. Finally, this buffer is copied again to a host visible one for checking.

### What happens?

On NV the first part works, the correct data arrives at the host side, but the altered data never arrives at the device side. On AMD not even the first part works and I already don't get the data on the host.

### The buffers:

- device_buffer is a buffer with VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT memory and VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_TRANSFER_SRC_BIT usage
- host_buffer is a buffer with VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT memory and VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_TRANSFER_SRC_BIT usage
- result_buffer is a buffer with VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT memory and VK_BUFFER_USAGE_TRANSFER_DST_BIT usage

**Comment by krOoze** [comment]
I get 404 on your link

**Comment by neXyon** [comment]
Sorry, fixed it!

**Comment by krOoze** [comment]
Your memory barrier have no access flags set.

**Comment by neXyon** [comment]
Do you mean they have the wrong access flags? Which ones are the correct ones?

**Comment by krOoze** [comment]
@neXyon Nevermind, just being blind.

**Comment by krOoze** [comment]
So, there is no host wait, and you hack around it by `vkGetEventStatus`.
I suppose that would account for execution dependency. What about the memory dependency?
There's probably more hackery needed to make it work (semaphore?).

**Comment by krOoze** [comment]
Host to device signalling should work normally.
Then again is not great either, because if you signal after the wait executes, then it is almost an undefined results (sic), according to [Queue Forward Progress](http://vulkan.hys.cz/1.1-extensions/vkspec.html#commandbuffers-submission-progress).

**Comment by neXyon** [comment]
### Device to host synchronization

For the memory dependency, I thought that the memory barrier before `vkCmdSetEvent` does that?

```c++
	barrier.srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
	barrier.dstAccessMask = VK_ACCESS_HOST_READ_BIT;
	vkCmdPipelineBarrier(command_buffer, VK_PIPELINE_STAGE_TRANSFER_BIT | VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT | VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
VK_PIPELINE_STAGE_HOST_BIT | VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT | VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT, 0, 1, &barrier, 0, nullptr, 0, nullptr);
```

I guess the question is, whether my busy wait with `vkGetEventStatus` is allowed. The relevant details from the specification read

> No command exists to wait for an event to become signaled on the host, but the current state of an event **can** be queried.

> If a `vkCmdSetEvent` or `vkCmdResetEvent` command is in a command buffer that is in the pending state, then the value returned by this command **may** immediately be out of date.

Neither of these statements explicitly states that this is either valid or not, it would be nice if the specification could be clear about that.

How would you do more hackery with a semaphore? The way that definitely works would be using a fence instead of an event, I guess? Then I would need two command buffers and I need two calls to vkQueueSubmit, the first one having the fence and the second one waiting for the even at the beginning.

### Host to device synchronization

With undefined results (sic), according to Queue Forward Progress, do you mean:

> If the event is not signaled within these limits, results are undefined and may include device loss.

and the case that these limits would be reached? Maybe especially if a fence is used? Would it then be better to wait for the fence and after the host code was executed submit the second command buffer instead of having it wait for an event?

**Comment by krOoze** [comment]
> For the memory dependency, I thought that the memory barrier before `vkCmdSetEvent` does that?

Does not, because it is async with `vkCmdSetEvent`. It would work if `vkCmdSetEvent::stageMask = VK_PIPELINE_STAGE_HOST_BIT`, alas that is not valid usage.

> I guess the question is, whether my busy wait with `vkGetEventStatus` is allowed.

It is allowed. But it is a matter of taking a hint from the API design.

I would expect, on most implementation, the Event host-side state to be cached, and only changed if the driver can prove the command buffer finished execution. Meaning it would piggyback on your waits on fences (and perhaps having its own internal fences if the impl likes you). But I would not expect mid-cmdbuffer device2host communication on most implementations.

That being said, I would half-expect you to wait forever in the loop (or at least a long time, until _something_ figures out the cmdbuffer ended executing).

> With undefined results (sic), according to Queue Forward Progress, do you mean:

Indeed. Considering I as a user do not know the nature of this "limit" that makes such API use a quantum superposition of being valid and invalid.

**Comment by krOoze** [comment]
N.b. `vkCmdPipelineBarrier` synchronisation scopes do not even include host-side access. `VK_PIPELINE_STAGE_HOST_BIT` is potentially allowed there only for dependency chaining purposes...

**Comment by neXyon** [comment]
Thanks a lot that clears some things up! It would be nice if the specification could state this more clearly!

Two of my questions remain still: Did you mean a fence instead of a semaphore? And if I use a fence would it just be better to submit the second command buffer after the fence, instead of having an event?

**Comment by krOoze** [comment]
@neXyon I was just thinking aloud with the semaphore. Mostly thought how to hack around and break things (as one shouldn't)...

But yeah, Fences for device-to-host communication.

**Comment by neXyon** [comment]
Okay, thanks. I guess the answer to my last question is that it depends. You might be able to get just a slightly lower latency with using an event, but at the cost of risking to run into the limit. So it has to be evaluated on an application- and device-dependent basis.

The conclusion now is that Device to Host synchronization is not possible with events. Should we leave this bug report open in hope that the specification will be updated to more explicitly mention this, or close the report?

**Comment by nvpbrown** [comment]
I just looked at this briefly when this issue was being reviewed and didn't read this super carefully.

Are you inserting a barrier in the command buffer that is executed *after* the command where the device waits on the event signaled by the host?  My first thought is that the host logic here is updating values that may be cached on the GPU/device with nothing to invalidate those caches.

**Comment by neXyon** [comment]
Yes, there is a barrier that I'm supplying to `vkCmdWaitEvents`:

```
	barrier.srcAccessMask = VK_ACCESS_HOST_WRITE_BIT;
	barrier.dstAccessMask = VK_ACCESS_TRANSFER_READ_BIT;

	// wait for host to device sync event

	vkCmdWaitEvents(command_buffer, 1, &host_to_device_sync_event, VK_PIPELINE_STAGE_HOST_BIT, VK_PIPELINE_STAGE_TRANSFER_BIT, 1, &barrier, 0, nullptr, 0, nullptr);
```

Is there another way to invalidate the device caches?

**Comment by nvpbrown** [comment]
@neXyon Thanks for the detail.  It does look like that barrier is attempting to do the right thing w.r.t. GPU cache hierarchies.

I haven't done all that much with Vulkan to date, so would need to look into this further with colleagues at NVIDIA with more experience.

**Comment by neXyon** [comment]
@nvpbrown That would be nice, thanks!

**Comment by krOoze** [comment]
> @neXyon Thanks for the detail. It does look like that barrier is attempting to do the right thing w.r.t. GPU cache hierarchies.

From my understanding of the spec I do see specific issues with it.

Firstly, when `vkCmdPipelineBarrier` talks about synchronization scopes, it talks only about commands with submission order:

> If `vkCmdPipelineBarrier` was recorded outside a render pass instance, the first synchronization scope includes all commands that occur earlier in submission order.

 In contrast with `vkCmdWaitEvents`, which expands its synchronization scopes in case of host stage:

> Event signal operations performed by `vkSetEvent` are only included in the first synchronization scope if `VK_PIPELINE_STAGE_HOST_BIT` is included in `srcStageMask`.

Secondly, assuming the `vkCmdPipelineBarrier` does properly define memory dependency between the transfer and host, the `vkCmdSetEvent` does still only have execution dependency on the transfer:

> When `vkCmdSetEvent` is submitted to a queue, it defines an execution dependency on commands that were submitted before it, and defines an event signal operation which sets the event to the signaled state.

Therefore I could be accessing the memory on the host before it is flushed from the device.

I am assuming `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` works somewhat same as doing manually `vkInvalidateMappedMemoryRanges` (could use some clarification). Which says:

`vkInvalidateMappedMemoryRanges` guarantees that device writes to the memory ranges described by `pMemoryRanges`, which have been made visible to the `VK_ACCESS_HOST_WRITE_BIT` and `VK_ACCESS_HOST_READ_BIT` access types, are made visible to the host.

As stated above, we do not know if the memory is in visible state at the point the `vkGetEventStatus` returns. We only know the **execution** of the transfer has finished.

**Comment by drakos-amd** [comment]
Fine-grained mid-command buffer synchronization between the host and the device should, in theory, be possible in the way you propose if my understanding it correct. Let's go through the individual steps:

> 1. A buffer is filled with increasing integers and uploaded to the device (device local buffer).
> 2. This buffer is copied to a host visible one.
> 3. The first event is set.

As already discussed above, if there's a device-to-host memory or execution dependency then the `vkCmdSetEvent` call should be preceeded by a `vkCmdPipelineBarrier` with its `dstStageMask` parameter set to the same value as the `stageMask` parameter specified in `vkCmdSetEvent` in order to ensure that you have the proper happens-before relation.

If there's a memory dependency then obviously `dstAccessMask` should include `VK_ACCESS_HOST_READ_BIT` and/or `VK_ACCESS_HOST_WRITE_BIT` (depending on what you do with the shared data, in your example it looks like only the latter is needed).

In the simplest case when there's no device-to-host memory or execution dependency then this stage mask could be simply `VK_PIPELINE_STAGE_TOP_OF_PIPE`, and no `vkCmdPipelineBarrier` call is needed, but I assume you have some dependency otherwise you wouldn't need the first event.

Also, if there's only an execution dependency but no memory dependency then you can simply specify the appropriate `stageMask` to signal-after directly to `vkCmdSetEvent` (no need for `vkCmdPipelineBarrier`).

(Note that there's no reason/need for accepting `VK_PIPELINE_STAGE_HOST_BIT` in `vkCmdSetEvent` contrarily to what has been suggested in a previous comment as `vkCmdSetEvent` specifies the pipeline stages after which the event should be set by the device, not the stage where the event will be subsequently waited for.)

> 4. The second event is waited for.

In order to have proper synchronization, the following parameters have to be specified to `vkCmdWaitEvents`:
- `srcStageMask` must be `VK_PIPELINE_STAGE_HOST_BIT`, as we're waiting for an event signaled by the host
- `srcAccessMask` of the memory barrier has to include `VK_ACCESS_HOST_WRITE_BIT`, as you're waiting for the host writes to become visible

The values of `dstStageMask` and `dstAccessMask` evidently will have to be set according to what you plan to do with the shared resource on the device after the wait.

> 5. Meanwhile the host waits for the first event.

Sure, spinning until `vkGetEventStatus` returns `VK_EVENT_SET` is fine.

> 6. After it has been set, it increments the numbers in the host visible buffer.

One thing to pay attention when reading/writing shared data in host-mapped allocations is that if the memory type of the allocation does not have the `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` property then you have to use `vkInvalidateMappedMemoryRanges` and/or `vkFlushMappedMemoryRanges` if you read and/or write the memory, respectively, to ensure that coherency between the host accesses and device accesses synchronized with `VK_ACCESS_HOST_READ_BIT`/`VK_ACCESS_HOST_WRITE_BIT`.

From your description it looks like you do use a memory type that has the `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` but I'd suggest you to give a try issuing an invalidate/flush for debugging purposes to see if it changes anything.

I also wonder whether using a memory type with or without `VK_MEMORY_PROPERTY_HOST_CACHED_BIT` changes anything.

> 7. Then it sets the second event.
> 8. The device then continues to copy the host visible buffer back to the device local buffer.

Considering the next operation you do with the buffer is a copy, it means the parameters of the `vkCmdWaitEvents` call in step 4 has to have:
- `dstStageMask` set to `VK_PIPELINE_STAGE_TRANSFER_BIT`
- `dstAccessMask` set to `VK_ACCESS_TRANSFER_READ_BIT`

The parameters you specified for `vkCmdWaitEvents` seem to match the above.

I did try out your example code on github on an AMD GPU and it seems to work just fine (no mismatches in the expected results).

In your example code the stage mask parameters of `vkCmdSetEvent` and its preceeding `vkCmdPipelineBarrier` are a bit superfluous, and in fact incorrect, as the `dstStageMask` parameter of `vkCmdPipelineBarrier` must be a strict superset of the `stageMask` parameter of `vkCmdSetEvent` as explained earlier, though in practice it happens to work as is (as implementations may map multiple API pipeline stages to the same underlying HW-specific value). The correct and minimal stage masks would be to:
- set the `srcStageMask` parameter of `vkCmdPipelineBarrier` to `VK_PIPELINE_STAGE_TRANSFER_BIT` (as you have a transfer preceeding the barrier), and
- set the `dstStageMask` parameter of `vkCmdPipelineBarrier` and the `stageMask` parameter of `vkCmdSetEvent` to the same value (e.g. `VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT`)

Indifferently though either with the above modifications or with the original (strictly speaking incorrect) parameters the test seem to work just fine on the configuration I've tested it on.

Can you provide more information on what AMD GPU, driver version, OS, and architecture (32-bit or 64-bit build) did you see the incorrect behavior on?

**Comment by neXyon** [comment]
Hi @drakos-amd,

thanks for your detailed reply!

I wasn't entirely sure about the stage masks of the `vkCmdSetEvent` and the `vkCmdPipelineBarrier`, that's why I set them so superfluous as you said. Actually, the `dstAccessMask` should probably be `VK_ACCESS_HOST_READ_BIT | VK_ACCESS_HOST_WRITE_BIT` since I'm reading and writing to the same buffer at the host. Setting it to that, I have to set the `dstStageMask` of the `vkCmdPipelineBarrier` to `VK_PIPELINE_STAGE_HOST_BIT` or I get a validation layer error, that the `dstAccessMask` is not supported. Following your comments I then have to set the `stageMask` of the `vkCmdSetEvent` to `VK_PIPELINE_STAGE_HOST_BIT` as well instead of `VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT`. I updated the gist accordingly. But isn't that wrong since the `stageMask` should be the synchronization scope before the event is set? I also tried `VK_PIPELINE_STAGE_HOST_BIT`, so it differed from the `dstStageMask`, but no different results.

I also tried `vkInvalidateMappedMemoryRanges` and/ `vkFlushMappedMemoryRanges` as well as `VK_MEMORY_PROPERTY_HOST_CACHED_BIT`, but it did not change anything, on my AMD setup I always get the "Wrong 1" and "Wrong 2" prints of my example. So here is the setup:

- AMD GPU: Radeon RX Vega 64
- Driver version: xf86-video-amdgpu 18.0.1 and vulkan-radeon 18.1.5
- OS: arch linux 64 bit (kernel: 4.17.13-arch1-1-ARCH)

**Comment by drakos-amd** [comment]
> Actually, the dstAccessMask should probably be VK_ACCESS_HOST_READ_BIT | VK_ACCESS_HOST_WRITE_BIT since I'm reading and writing to the same buffer at the host.

Sure.

> Setting it to that, I have to set the dstStageMask of the vkCmdPipelineBarrier to VK_PIPELINE_STAGE_HOST_BIT or I get a validation layer error, that the dstAccessMask is not supported.

You're right about that. My bad, as I only mentioned there the needed flags to sync `vkCmdPipelineBarrier` with `vkCmdSetEvent`, but didn't account for the `vkCmdPipelineBarrier` also needing the memory dependency with the host access.

> Following your comments I then have to set the stageMask of the vkCmdSetEvent to VK_PIPELINE_STAGE_HOST_BIT as well instead of VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT.

That's still not necessary though, as it is enough to only have the `stageMask` parameter of `vkCmdSetEvent` be a subset of the `dstStageMask` specified for `vkCmdPipelineBarrier` to ensure the proper dependency. I was not accurate in my original reply when I said they have to exactly match. The reason that's enough is that synchronization commands can establish happens-before relationships with multiple subsequent set of operations and with your fix to my original suggestion you practically establish two in this case:
- `VK_PIPELINE_STAGE_HOST_BIT` in the `vkCmdPipelineBarrier` ensures that the memory barrier happens-before the host access to the resource (assuming it's otherwise has the proper execution dependency, which is what you use the event for)
- `VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT` ensures that the memory barrier happens-before the event is set by `vkCmdSetEvent` also using `VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT` (thus also makes sure the above happens-before the host seeing the event being set)

I hope that makes sense.

> I also tried vkInvalidateMappedMemoryRanges and/ vkFlushMappedMemoryRanges as well as VK_MEMORY_PROPERTY_HOST_CACHED_BIT, but it did not change anything, on my AMD setup I always get the "Wrong 1" and "Wrong 2" prints of my example.

We'll look into testing this on the particular configuration you have. Thanks for the information.

**Comment by drakos-amd** [comment]
We verified our (AMD) driver implementation on the specific OS and HW configuration and the issue does not reproduce. As you're not using AMD's Linux driver (proprietary or open-source) but the Mesa RADV driver, you should contact them about the issue.

**Comment by krOoze** [comment]
@drakos-amd  Works for me as well on W10 with AMD. I still do not see how this would be legal.

Based on the above we define these dependencies:
 - `TRANSFER`&rarr;`HOST` memory dependency (there's not really any execution dependency, because `HOST` is a "fake" stage, and host reads are non-Q ops)
 - `TRANSFER`&rarr;`BOTTOM` execution dependency (with device-domain availability op)
 - `BOTTOM`&rarr;event-signal execution dependency
 - above event-signal&rarr;host-rw execution dependency
 - by Coherent every available memory in host domain is implicitly considered also visible

We never set a dependency from the cross-domain op to the event-signal. We **only** know the domain op happens-after `TRANSFER`, and the event-signal also happens-after `TRANSFER`. So AFAICT we have race-condition of those two. The host might be reading data that is not visible to the host domain yet.

At least that is my reading of the spec. May I have some quotes saying otherwise?

**Comment by neXyon** [comment]
@drakos-amd: Thanks! I just tried with amdvlk and there it indeed works! I'll report a bug to Mesa RADV.

@krOoze: It would be nice if this could be cleared up in the specification - either in a way that the specification supports this usage, allows it (with a way to check if it's possible; maybe an extension?), or disallows it. What would be the way forward here to get to one of these three points?

@nvpbrown: Have you had a chance to look into this with your colleagues?

**Comment by drakos-amd** [comment]
> We never set a dependency from the cross-domain op to the event-signal. We only know the domain op happens-after TRANSFER, and the event-signal also happens-after TRANSFER. So AFAICT we have race-condition of those two. The host might be reading data that is not available(nor visible) to the host domain yet.

I'm not sure I follow. Availability, visibility, and domain operations are part of the memory dependency and thus happen-before any operation in the stages specified in dstStageMask process subsequent operations, so vkCmdSetEvent and any host operations (due to the use of `VK_PIPELINE_STAGE_HOST_BIT`) should also happen after that.

Maybe the confusion comes from you separating the `TRANSFER` -> `HOST` and the `TRANSFER` -> `BOTTOM` dependencies, while in practice there's a `TRANSFER` -> { `HOST` + `BOTTOM` } dependency. When the spec defines execution dependencies it does define them for the sets of scopes (pipeline stages in this case), not individual pairs, even though a dependency between the sets also implies a dependency between individual pairs.

@krOoze, if you feel something is missing or otherwise under-specified in the spec please file a separate issue on that.

I'll not close this issue quite yet in case @nvpbrown has any comments from their implementation's perspective.

**Comment by krOoze** [comment]
@drakos-amd  I guess. I am instructed to think that way by the spec:

> For example, a synchronization command with multiple pipeline stages in its stage masks effectively generates one dependency between each source stage and each destination stage.

Might be just that `HOST_STAGE` is weird. Or there is some weird emergent relationship between stages masked together in `dstStage` that I did not consider.

Is this the code snippet that we settled on (which is supposed to be valid, and should work correctly)?:

```C++
// VkResult ch( VkResult r ){ if( r >= 0 ) return r; else throw "err";  }
// std::vector<uint8_t> test_data = ...;
// VkDeviceMemory buffer_memory  = ...;
// VkBuffer buffer = ... bind buffer_memory;
// VkDevice device = ...;
// VkQueue q = ...;
// VkEvent  d2h_event = ...;
// VkCommandBuffer cmdbuff = ...;
void* mapped_buffer;
ch( vkMapMemory( device,  buffer_memory, 0, VK_WHOLE_SIZE, 0, &mapped_buffer ) );

const VkCommandBufferBeginInfo bi = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO };
ch( vkBeginCommandBuffer( cmdbuff, &bi ); )
    vkCmdUpdateBuffer( cmdbuff, buffer, 0, test_data.size(), test_data.data() );

    VkBufferMemoryBarrier buff_barrier = {};
        buff_barrier.sType = VK_STRUCTURE_TYPE_BUFFER_MEMORY_BARRIER;
        buff_barrier.srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
        buff_barrier.dstAccessMask = VK_ACCESS_HOST_READ_BIT | VK_ACCESS_HOST_WRITE_BIT ;
        buff_barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
        buff_barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
        buff_barrier.buffer = buffer;
        buff_barrier.offset = 0;
        buff_barrier.size = VK_WHOLE_SIZE;
    vkCmdPipelineBarrier(
        cmdbuff,
        VK_PIPELINE_STAGE_TRANSFER_BIT,
        VK_PIPELINE_STAGE_HOST_BIT | VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
        0,
        0, nullptr, 1, &buff_barrier, 0, nullptr
    );
    vkCmdSetEvent( cmdbuff, d2h_event, VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT );
ch( vkEndCommandBuffer( cmdbuff ); );

VkSubmitInfo si = {};
    si.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
    si.commandBufferCount = 1;
    si.pCommandBuffers = &cmdbuff;
ch( vkQueueSubmit( q, 1, &si, VK_NULL_HANDLE ) );

while(  ch( vkGetEventStatus( device, d2h_event ) ) != VK_EVENT_SET  ) continue;

for( size_t i = 0; i < test_data.size; ++i ){
    assert( reinterpret_cast<uint8_t>(mapped_buffer)[i] == test_data[i] );
}
```

> Availability, visibility, and domain operations are part of the memory dependency and thus happen-before any operation in the stages specified in dstStageMask process subsequent operations

**Comment by krOoze** [comment]
BTW `vkCmdWaitEvents` before `vkSetEvent` was made invalid usage now in 1.1.104...

**Comment by neXyon** [comment]
Interesting, it now states:

> The event **must** be set before the `vkCmdWaitEvents` command is executed.

This sounds weird. Does it simply mean that the execution of `vkCmdWaitEvents` waits until the event is set (which it of course **must** do, that's the whole point of `vkCmdWaitEvents`) or does it mean that the event must be set before the execution even reaches the `vkCmdWaitEvents` command, i.e. rendering that use case pointless, since you then could simply skip the wait for that event. Or am I not nuanced enough with this distinction?

**Comment by Tobski** [comment]
> Interesting, it now states:

> >    The event must be set before the vkCmdWaitEvents command is executed.

> This sounds weird. Does it simply mean that the execution of vkCmdWaitEvents waits until the event is set (which it of course must do, that's the whole point of vkCmdWaitEvents) or does it mean that the event must be set before the execution even reaches the vkCmdWaitEvents command, i.e. rendering that use case pointless, since you then could simply skip the wait for that event. Or am I not nuanced enough with this distinction?

So this change basically made host setting almost useless, and yes, it was done deliberately. Realistically this has always been hideously undefined - an implementation that times out the wait with a timeout of zero has always been allowed. Rather than keeping up the facade that this actually works, we've now made it undefined, but noted that implementation-specific knowledge might allow this to work in some circumstances - but it's outside the scope of the Vulkan spec to define that.

I think that basically means the example posited at the top of this issue is actually now invalid...?

**Comment by neXyon** [comment]
To summarize: as far as I understand now, from the specs:

> Events are a synchronization primitive that can be used to insert a fine-grained dependency between commands submitted to the same queue, or between the host and a queue.

Only the first part is actually practical (synchronization on the same queue) and the second part is not. Neither device to host synchronization (mostly this issue) nor host to device synchronization (these recent changes) can be trusted to work reliably? So the only way to do these kinds of synchronization is via fences (device to host) and queue submission (host to device)?

**Comment by procedural** [comment]
> Neither device to host synchronization (mostly this issue) nor host to device synchronization (these recent changes) can be trusted to work reliably?

Correct me if I'm wrong, this is how I interpret these changes:

Device to host: `vkGetEventStatus`, should still work fine, I think?

Host to device: `vkSetEvent`, this is now fully platform dependent, if a platform allows to wait on a device for some time for an event to become signaled from the host, then you can you can use it.

**Comment by procedural** [comment]
@Tobski

> an implementation that times out the wait with a timeout of zero has always been allowed

Is there any way to query this timeout value?

**Comment by Tobski** [comment]
> Only the first part is actually practical (synchronization on the same queue) and the second part is not. Neither device to host synchronization (mostly this issue) nor host to device synchronization (these recent changes) can be trusted to work reliably? So the only way to do these kinds of synchronization is via fences (device to host) and queue submission (host to device)?

This is pretty much correct. Synchronizing device to host with events is still practical in some circumstances, and there's actually a super limited set of things you can do with host signalling of events, but in general applications should be using the more robustly defined fences and semaphores.

**Comment by Tobski** [comment]
@procedural 

> Device to host: vkGetEventStatus, should still work fine, I think?

Yep.

> Host to device: vkSetEvent, this is now fully platform dependent, if a platform allows to wait on a device for some time for an event to become signaled from the host, then you can use it.

Practically yes, though from the Vulkan PoV it is completely invalid, and we'd expect a validation error to show up at some point. Someone would need to define an extension to make this work correctly. We really should have done that from the start.

> Is there any way to query this timeout value?

There is not - from the Vulkan PoV it is now completely undefined. As above, someone would need to define an extension to make this work.

**Comment by Tobski** [comment]
Yes, you can busy wait on vkGetEventStatus if you want. Nothing in the spec forbids this.

**Comment by neXyon** [comment]
Well, the whole point of this issue was to get a more precise specification of how event synchronization between host and device works. But since kinda the opposite happened, I guess I can't expect that to happen. So the issue can be closed?

**Comment by plasmacel** [comment]
> 

I don't think that this is correct. `dstStageMask = BOTTOM_OF_PIPE_BIT`, which is equivalent to `VK_PIPELINE_STAGE_NONE` doesn't form an execution dependency chain with the succeeding event set operations or semaphore signals.

**Comment by krOoze** [comment]
@plasmacel It's a long thread, so not sure what precisely you are responding to.

It is more of other way around. `VK_PIPELINE_STAGE_NONE` is new equivalent to `dstStageMask = BOTTOM_OF_PIPE_BIT` or `srcStageMask = TOP_OF_PIPE_BIT`. Those do participate in dependency chains, especially things like sempahores and fences.
